test/php/gutenberg-coding-standards/Gutenberg/Sniffs/CodeAnalysis/ForbiddenFunctionsAndClassesSniff.php

```diff
@@ -0,0 +1,231 @@
+			return;
+		}
+
+		$tokens     = $phpcs_file->getTokens();
+		$class_name = $tokens[ $stack_pointer ]['content'];
+
+		foreach ( $this->forbidden_classes as $forbidden_class ) {
+			$regexp = sprintf( '/^%s$/', $forbidden_class );
+			if ( 1 !== preg_match( $regexp, $class_name ) ) {
+				// The class has a valid name; bypassing further checks.
+				return;
+			}
+		}
+
+		$error_message = 'It\'s not allowed to use the "' . $class_name . '" class as its name matches the forbidden pattern: "' . $regexp . '".';
+		$phpcs_file->addError( $error_message, $stack_pointer, 'ForbiddenClassUsage' );
+	}
+
+	/**
+	 * Checks for forbidden function usage.
+	 *
+	 * @param File $phpcs_file    File being scanned.
+	 * @param int  $stack_pointer Position of the text token in the token stack.
+	 */
+	private function check_function_usage( File $phpcs_file, $stack_pointer ) {
+		if ( empty( $this->forbidden_functions ) ) {
+			// Nothing to process.
+			return;
+		}
+
+		if ( $this->check_if_token_guarded( $phpcs_file, $stack_pointer ) ) {
+			// The token is guarded. Nothing to process.
+			return;
+		}
+
+		$tokens        = $phpcs_file->getTokens();
+		$function_name = $tokens[ $stack_pointer ]['content'];
+
+		foreach ( $this->forbidden_functions as $disallowed_function_call ) {
+			$regexp = sprintf( '/^%s$/', $disallowed_function_call );
+			if ( 1 !== preg_match( $regexp, $function_name ) ) {
+				// The function has a valid name; bypassing further checks.
+				return;
+			}
+		}
+
+		$error_message = 'It\'s not allowed to call the "' . $function_name . '()" function as its name matches the forbidden pattern: "' . $regexp . '".';
+		$phpcs_file->addError( $error_message, $stack_pointer, 'ForbiddenFunctionCall' );
+	}
+
+	/**
+	 * This function looks for a specific if condition that wraps the token
+	 * to determine if it's protected from execution outside of Gutenberg.
+	 *
+	 * Example of a guarded function (works similarly for classes):
+	 * if ( defined( 'IS_GUTENBERG_PLUGIN' ) && IS_GUTENBERG_PLUGIN ) {
+	 *     // gutenberg_prefixed_i_am_allowed_to_be_called_since_i_am_guarded_function();
+	 * }
+	 *
+	 * @param File $phpcs_file    File being scanned.
+	 * @param int  $stack_pointer Position of the text token in the token stack.
+	 *
+	 * @return bool               Returns true if the token is guarded.
+	 */
+	private function check_if_token_guarded( File $phpcs_file, $stack_pointer ) {
+		if ( false === $phpcs_file->hasCondition( $stack_pointer, T_IF ) ) {
+			// No wrapping IF tokens. Stop processing.
+			return false;
+		}
+
+		$tokens = $phpcs_file->getTokens();
+
+		// It has to start from the matched condition closest to the passed token.
+		$conditions = array_reverse( $tokens[ $stack_pointer ]['conditions'], true );
+
+		// Matches the "if ( defined( 'IS_GUTENBERG_PLUGIN' ) && IS_GUTENBERG_PLUGIN" string.
+		$regexp = '/if\s*\(\s*defined\(\s*(\'|")IS_GUTENBERG_PLUGIN(\'|")\s*\)\s*&&\s*IS_GUTENBERG_PLUGIN/';
```

Comment:
I'm wondering if there would ever be a reason to make this check configurable? Checking for `IS_GUTENBERG_PLUGIN` makes sense in block_library or any other PHP files that may be auto-generated in core from the npm packages. I guess this kind of logic wouldn't be needed at all if we were using the sniff in plugin-specific PHP files.

Reply:
The sniff is currently only checking the `/packages/block-library` folder: https://github.com/WordPress/gutenberg/pull/57738/files#diff-05ae9cddcaec1e845771a7db224961439f83ef5939ec67d3a48744cb34d7e58bR157
So, it doesn't check files in other folders.
I hope I've understood your question correctly.